src/amb_kf.c

```diff
@@ -162,21 +190,100 @@ static void incorporate_scalar_measurement(u32 state_dim, double *h, double R,
   DEBUG_EXIT();
 }
 
+/** Get the weighted sum of squared innovations.
+ * It's a normalized error metric. High is bad.
+ * More precisely, we square the difference between the predicted observations
+ * and the actual observations, divide them by their variances (FPF' + R),
+ * but pretending they're independent, and then sum them.
+ * It's: sum_i (z - H * x)_i / (F * P * F' + R)_ii.
+ * If FPF'+R was actually diagonal and these measurements were independent,
+ * these would be chi square with kf->obs_dim degrees of freedom.
+ *
+ * \param kf        Kalman filter struct
+ * \param decor_obs Decorrelated observation vector
+ * \return          The weighted SOS.
+ */
+double get_sos_innov(const nkf_t *kf, const double *decor_obs)
+{
+  double predicted_obs[kf->obs_dim];
+  matrix_multiply(kf->obs_dim, kf->state_dim, 1,
+                  kf->decor_obs_mtx, kf->state_mean, predicted_obs);
+  double hu[kf->obs_dim * kf->state_dim];
+  /* TODO use the fact that U is unit triangular to save a ton of time */
+  matrix_multiply(kf->obs_dim, kf->state_dim, kf->state_dim,
+                  kf->decor_obs_mtx, kf->state_cov_U, hu);
+  /* (H * U * D * U^T * H^T)_ii = (HU * D * HU^T)_ii
+   *                            = Sum_kl (HU_ik * D_kl * HU^T_li)
+   *                            = Sum_kl (HU_ik * D_kl * HU_il)
+   *                            = Sum_k (HU_ik * D_kk * HU_ik) */
```

Comment:
Seems like there is a bit of inconsistency in notation - we use F in some places and H in others? I think this comment should add in the step before where we go from `(F * P * F' + R)_ii` to `(H * U * D * U^T * H^T)_ii`.